### PR TITLE
feat(ec2): add support for al2022 and amzn2 with kernel 5.x

### PR DESCRIPTION
**Pull request by robertd**
This PR is aimed to expand CDK ability to support `al2022` and `amzn2` with kernel 5.x

```
"/aws/service/ami-amazon-linux-latest/al2022-ami-kernel-5.10-arm64"
"/aws/service/ami-amazon-linux-latest/al2022-ami-minimal-kernel-5.10-arm64"
"/aws/service/ami-amazon-linux-latest/amzn-ami-hvm-x86_64-ebs"
"/aws/service/ami-amazon-linux-latest/amzn-ami-hvm-x86_64-gp2"
"/aws/service/ami-amazon-linux-latest/amzn-ami-hvm-x86_64-s3"
"/aws/service/ami-amazon-linux-latest/amzn-ami-minimal-hvm-x86_64-s3"
"/aws/service/ami-amazon-linux-latest/amzn2-ami-hvm-arm64-gp2"
"/aws/service/ami-amazon-linux-latest/amzn2-ami-hvm-x86_64-ebs"
"/aws/service/ami-amazon-linux-latest/amzn2-ami-hvm-x86_64-gp2"
"/aws/service/ami-amazon-linux-latest/amzn2-ami-kernel-5.10-hvm-x86_64-ebs"
"/aws/service/ami-amazon-linux-latest/al2022-ami-kernel-5.10-x86_64"
"/aws/service/ami-amazon-linux-latest/al2022-ami-minimal-kernel-5.10-x86_64"
"/aws/service/ami-amazon-linux-latest/amzn-ami-minimal-hvm-x86_64-ebs"
"/aws/service/ami-amazon-linux-latest/amzn-ami-minimal-pv-x86_64-ebs"
"/aws/service/ami-amazon-linux-latest/amzn-ami-minimal-pv-x86_64-s3"
"/aws/service/ami-amazon-linux-latest/amzn-ami-pv-x86_64-ebs"
"/aws/service/ami-amazon-linux-latest/amzn-ami-pv-x86_64-s3"
"/aws/service/ami-amazon-linux-latest/amzn2-ami-kernel-5.10-hvm-arm64-gp2"
"/aws/service/ami-amazon-linux-latest/amzn2-ami-kernel-5.10-hvm-x86_64-gp2"
"/aws/service/ami-amazon-linux-latest/amzn2-ami-minimal-hvm-arm64-ebs"
"/aws/service/ami-amazon-linux-latest/amzn2-ami-minimal-hvm-x86_64-ebs"
```

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by robertd** [comment]
@njlynch I'd really appreciate your feedback when you get a chance. I still think this needs to be refactored. I'm not really happy with cleanliness of this PR. `al2022` and `amzn2` (kernel-5.10) ssm param entries are a bit different from other entries... i.e. missing storage and virtualization components, etc.

Just let me know what you think and I'll follow your lead. In the meantime, I've converted this PR to a draft.

File: packages/@aws-cdk/aws-ec2/lib/machine-image.ts [hunk 1]
```diff
@@ -468,7 +499,7 @@ export enum AmazonLinuxStorage {
   /**
    * S3-backed storage
    */
-  S3 = 'ebs',
+  S3 = 's3',
```
Comment (robertd) [inline packages/@aws-cdk/aws-ec2/lib/machine-image.ts:502]: @njlynch Was this an omission... or by design?

**Comment by robertd** [comment]
Also, I'm planning on updating `Readme.MD` and probably adding more tests to cover all possible ssm linux parameters coming from `/aws/service/ami-amazon-linux-latest` path.

File: packages/@aws-cdk/aws-ec2/lib/machine-image.ts (hunk 1 above)
Comment (njlynch) [reply packages/@aws-cdk/aws-ec2/lib/machine-image.ts:502]: I imagine this was a mistake. Somehow it's gone unnoticed for 3 years!

File: packages/@aws-cdk/aws-ec2/lib/machine-image.ts [hunk 2]
```diff
@@ -374,15 +381,24 @@ export class AmazonLinuxImage extends GenericSSMParameterImage {
    * Return the SSM parameter name that will contain the Amazon Linux image with the given attributes
    */
   public static ssmParameterName(props: AmazonLinuxImageProps = {}) {
-    const generation = (props && props.generation) || AmazonLinuxGeneration.AMAZON_LINUX;
-    const edition = (props && props.edition) || AmazonLinuxEdition.STANDARD;
-    const virtualization = (props && props.virtualization) || AmazonLinuxVirt.HVM;
-    const storage = (props && props.storage) || AmazonLinuxStorage.GENERAL_PURPOSE;
-    const cpu = (props && props.cpuType) || AmazonLinuxCpuType.X86_64;
+    let generation = (props && props.generation) || AmazonLinuxGeneration.AMAZON_LINUX;
```
Comment (njlynch) [inline packages/@aws-cdk/aws-ec2/lib/machine-image.ts:384]: I'd prefer we keep the `const`s where possible.
```suggestion
    const generation = (props && props.generation) || AmazonLinuxGeneration.AMAZON_LINUX;
```

File: packages/@aws-cdk/aws-ec2/lib/machine-image.ts [hunk 3]
```diff
@@ -374,15 +381,24 @@ export class AmazonLinuxImage extends GenericSSMParameterImage {
...
+    let edition = (props && props.edition) || AmazonLinuxEdition.STANDARD;
+    let kernel = (props && props.kernel) || undefined;
+    let virtualization: AmazonLinuxVirt | undefined = (props && props.virtualization) || AmazonLinuxVirt.HVM;
+    let cpu = (props && props.cpuType) || AmazonLinuxCpuType.X86_64;
+    let storage: AmazonLinuxStorage | undefined = (props && props.storage) || AmazonLinuxStorage.GENERAL_PURPOSE;
+
+    if (generation === AmazonLinuxGeneration.AMAZON_LINUX_2022) {
+      kernel = AmazonLinuxKernel.KERNEL5_X;
+      storage = undefined;
+      virtualization = undefined;
```
Comment (njlynch) [inline packages/@aws-cdk/aws-ec2/lib/machine-image.ts:394]: Setting these to undefined is one approach; the other would be to explicitly throw an error if the user has set them. The latter might yield less surprises down the road, when someone in the future is trying to use the hypothetical EBS-storage or PV-virtualized 2022 AL, but isn't getting the results they expect.

**Review by njlynch — CHANGES_REQUESTED** [review]
I think it looks good.

Re: a refactoring of sorts. I could potentially see creating an AmazonLinux2022Image class with less properties, but I'm not sure that's the right model. I think what you have makes as good sense as any a natural evolution of the existing model.

File: packages/@aws-cdk/aws-ec2/lib/machine-image.ts (hunk 1 above)
Comment (robertd) [reply packages/@aws-cdk/aws-ec2/lib/machine-image.ts:502]: 😅

File: packages/@aws-cdk/aws-ec2/lib/machine-image.ts (hunk 3 above)
Comment (robertd) [reply packages/@aws-cdk/aws-ec2/lib/machine-image.ts:394]:
```ts
let virtualization: AmazonLinuxVirt | undefined = (props && props.virtualization) || AmazonLinuxVirt.HVM;
...
let storage: AmazonLinuxStorage | undefined = (props && props.storage) || AmazonLinuxStorage.GENERAL_PURPOSE;
```

@njlynch  I like the approach of throwing an error. However, both `virtualization` and `storage` are set to `HVM` and `GENERAL_PURPOSE` above, thus always triggering the error when AL2022 is selected. That's the reason why I went with `undefined` approach for these two variables.  Are there any other ways to make the error throw approach happen?

Comment (njlynch) [reply packages/@aws-cdk/aws-ec2/lib/machine-image.ts:394]: Something like this?

```
    let virtualization: AmazonLinuxVirt | undefined;
    let storage: AmazonLinuxStorage | undefined;

    if (generation === AmazonLinuxGeneration.AMAZON_LINUX_2022) {
      kernel = AmazonLinuxKernel.KERNEL5_X;
      if (props && props.storage) { } // throw error
      if (props && props.virtualization) {  } // throw error
    } else {
      storage = (props && props.storage) || AmazonLinuxStorage.GENERAL_PURPOSE;
      virtualization = (props && props.virtualization) || AmazonLinuxVirt.HVM;
    }
```

Comment (robertd) [reply packages/@aws-cdk/aws-ec2/lib/machine-image.ts:394]: Thank you!

**Comment by robertd** [comment]
@njlynch let me know if you'd like to see more integration tests or other tests in general. Thanks!

**Review by njlynch — APPROVED** [review]
Great!

**Pull request merged**